### PR TITLE
add go-counting

### PR DESCRIPTION
**Pull request by Average-user**

### CODE REVIEW

**Comment by Average-user** [comment]
@razum2um 
@haus 

I was wandering if it would be better to use, the tests cases and conventions generated by https://github.com/razum2um/exercism-generator

```Clojure
(ns go-counting-test
  (:require [clojure.test :refer [deftest is testing]]
            [go-counting :refer [territory]]))

(deftest black-corner-territory-on-5x-5-board
  (testing "Black corner territory on 5x5 board"
    (is (= {:territory [[0 0] [0 1] [1 0]], :owner "BLACK"}
           (territory ["  B  " " B B " "B W B" " W W " "  W  "] 0 1)))))

(deftest white-center-territory-on-5x-5-board
  (testing "White center territory on 5x5 board"
    (is (= {:territory [[2 3]], :owner "WHITE"}
           (territory ["  B  " " B B " "B W B" " W W " "  W  "] 2 3)))))

(deftest open-corner-territory-on-5x-5-board
  (testing "Open corner territory on 5x5 board"
    (is (= {:territory [[0 3] [0 4] [1 4]], :owner "NONE"}
           (territory ["  B  " " B B " "B W B" " W W " "  W  "] 1 4)))))

(deftest a-stone-and-not-a-territory-on-5x-5-board
  (testing "A stone and not a territory on 5x5 board"
    (is (= {:territory [], :owner "NONE"}
           (territory ["  B  " " B B " "B W B" " W W " "  W  "] 1 1)))))

(deftest invalid-because-x-is-too-low-for-5x-5-board
  (testing "Invalid because X is too low for 5x5 board"
    (is (thrown? Throwable
                 (territory ["  B  " " B B " "B W B" " W W " "  W  "] -1 1)))))

(deftest invalid-because-x-is-too-high-for-5x-5-board
  (testing "Invalid because X is too high for 5x5 board"
    (is (thrown? Throwable
                 (territory ["  B  " " B B " "B W B" " W W " "  W  "] 5 1)))))

(deftest invalid-because-y-is-too-low-for-5x-5-board
  (testing "Invalid because Y is too low for 5x5 board"
    (is (thrown? Throwable
                 (territory ["  B  " " B B " "B W B" " W W " "  W  "] 1 -1)))))

(deftest invalid-because-y-is-too-high-for-5x-5-board
  (testing "Invalid because Y is too high for 5x5 board"
    (is (thrown? Throwable
                 (territory ["  B  " " B B " "B W B" " W W " "  W  "] 1 5)))))

(deftest one-territory-is-the-whole-board
  (testing "One territory is the whole board"
    (is (= {:territory-white [], :territory-black [], :territory-none [[0 0]]}
           (territories [" "])))))

(deftest two-territory-rectangular-board
  (testing "Two territory rectangular board"
    (is (= {:territory-white [[3 0] [3 1]],
            :territory-black [[0 0] [0 1]],
            :territory-none []}
           (territories [" BW " " BW "])))))

(deftest two-region-rectangular-board
  (testing "Two region rectangular board"
    (is (= {:territory-white [],
            :territory-black [[0 0] [2 0]],
            :territory-none []}
           (territories [" B "])))))
```
I don't like territory outputs being vectors and not sets though.

**Comment by razum2um** [comment]
thanks for reminder 👍 
@haus actually,  I thought to fully regenerate clojure test suites for all exercises

**Comment by Average-user** [comment]
@razum2um It actually has a bug (your program), I think. `[go-counting :refer [territory]]` It does not refer `territories`

**Comment by Average-user** [comment]
I think the option to modify whats generated should always exist though.

**Comment by haus** [comment]
Thanks for the PR @Average-user ! I'll try to look this over in the next couple of days.

**Comment by razum2um** [comment]
@Average-user I fixed multiple requires https://github.com/razum2um/exercism-generator/commit/ee09d165dd255b5b652de0655fd19318b46407b6

Regarding the rest:
- I could extract arguments with a common name and value to the top `def`s
- I see no clear way how to tell zprint to format vector like board multiline notwithstanding all of it suits into 1 line. perhaps this may be possible if it would have something like `option-fn-first` but passing the whole vector (and warn about performance penalty)
- And I see no way to guess when a vector should be a set :( Such json, let's rewrite canonical data with something what preserves types (joking)

File: exercises/go-counting/src/example.clj [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+(defn neighbors [f stones [x y]]
+  (->> [[0 1] [0 -1] [1 0] [-1 0]]
+       (map (fn [[j k]] [(+ x j) (+ y k)]))
+       (filter (comp f stones))))
+
+(defn territory-of [stones [x y]]
+  (if (= :free (stones [x y]))
+    (letfn [(f [[seen fronteir]]
+              (as-> (reduce conj seen fronteir) nseen
```
Comment (haus) [inline exercises/go-counting/src/example.clj]: It looks like this `as->` could be replaced with a `let` binding?

File: exercises/go-counting/src/example.clj [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+(defn grid->graph [xs]
+  (->> (for [x (-> xs count range), y (-> xs first count range)]
+         (case (get-in xs [x y])
+           \W     [[y x] :white]
+           \B     [[y x] :black]
+           \space [[y x] :free]))
+       (into {})))
+
+(defn neighbors [f stones [x y]]
```
Comment (haus) [inline exercises/go-counting/src/example.clj]: Something slightly more meaningful than `f` might be good here? It only seems used on a set of owners, so maybe `owners` would be a good name.

File: exercises/go-counting/src/example.clj [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+(defn territory [grid [x y]]
+  (let [stones    (grid->graph grid)
+        territory (territory-of stones [x y])]
+    (if (nil? (stones [x y]))
+      (throw (Throwable. "Invalid coordinate!"))
+      {:stones territory :owner (territory-owner stones territory)})))
+
+(defn territories [grid]
+  (let [territories (->> grid grid->graph keys (map (partial territory grid)))
+        f #(->> territories (filter (comp % :owner)) (map :stones) (reduce concat) set)]
```
Comment (haus) [inline exercises/go-counting/src/example.clj]: Similar comment to above about a more meaningful name than `f`. Maybe `find-territory-for` or something like that?

File: exercises/go-counting/src/example.clj (hunk 2 above)
Comment (Average-user) [reply exercises/go-counting/src/example.clj]: I don't know why yesterday the comments wouldn't  appear, so I send many

**Comment by haus** [comment]
Awesome @Average-user . Thanks!

**Pull request merged**